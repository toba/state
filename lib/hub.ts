import { merge, is, removeItem } from '@toba/utility';
import { StateComponent } from '../jsx/stateful';
//import { statusPrefix } from "lib/i18n/language";
//import text from "lib/i18n/localize";
//import log from "lib/logger";

/**
 * Method for stores or simple components to handle actions dispatched from a
 * component.
 */
export type StateHandler = (action: number, data?: any) => void;

/**
 * Method for components to handle update dispatched from store.
 */
export type ViewHandler = () => void;

export interface State {
   [key: string]: any;
}

/**
 * Dispatcher in the standard Flux pattern. It subscribes stores and stateful
 * components to receive events from components.
 *
 * https://github.com/facebook/flux/tree/master/examples/flux-concepts
 */
export const flux = {
   /**
    * Global registry of state stores or stateful components that should receive
    * component action messages.
    */
   _handlers: [] as StateHandler[],

   /**
    * Dispatch action to every subscribed state store or component. If `data`
    * will be relayed to a service call then the service must define a matching
    * `struct` to unmarshall the JSON.
    */
   emit(action: number, data?: any) {
      this._handlers.forEach(fn => {
         fn(action, data);
      });
   },

   /**
    * Subscribe store or component to receive actions from components or other
    * stores.
    */
   subscribe<S extends StateStore<any> | StateComponent<any, any>>(
      stateful: S
   ): S {
      // bind creates a new function which must be reassigned if it's later to
      // be matched by remove()
      stateful.handler = stateful.handler.bind(stateful);
      this._handlers.push(stateful.handler);
      return stateful;
   },

   /**
    * Remove store or component subscription. This will usually only be used by
    * dismounting components while stores remain active for the life of the
    * application.
    */
   remove<S extends StateStore<any> | StateComponent<any, any>>(stateful: S) {
      removeItem(this._handlers, stateful.handler);
      return this;
   }
};

/**
 * State stores maintain and modify state based on dispatched action messages
 * then notify subscribed components when state has changed. Stores may call
 * services with Service IDs to update their state.
 *
 * Create as many state stores as are useful for avoiding duplicity and
 * complexity.
 */
export class StateStore<S extends State> {
   state: S;
   _initial: S;
   _handlers: ViewHandler[] = [];

   /**
    * Construct store with its initial state. If `reset()` is called, it will
    * return to this state.
    */
   constructor(initial: S) {
      this._initial = initial;
      this.state = merge(this._initial);
   }

   /**
    * Revert to initial state.
    */
   reset(): S {
      this.state = merge(this._initial);
      return this.state;
   }

   /**
    * Merge new with existing state and optionally emit change event.
    */
   update<K extends keyof S, P>(
      values:
         | ((prevState: Readonly<S>, props: P) => Pick<S, K> | S)
         | (Pick<S, K> | S),
      emitChange = true
   ) {
      this.state = merge(this.state, values);
      if (emitChange) {
         this.changed();
      }
   }

   /**
    * Return current state, optionally resetting to initial value.
    */
   load(force = false): S {
      return force || is.empty(this.state) ? this.reset() : this.state;
   }

   /**
    * Add a view handler to be notified when state changes.
    */
   subscribe(fn: ViewHandler) {
      if (fn != null) {
         this._handlers.push(fn);
      }
   }

   /**
    * Override to handle messages sent from components.
    */
   handler<T>(_action: number, _data?: T) {
      return;
   }

   /**
    * Remove a view handler.
    */
   remove(fn: ViewHandler) {
      removeItem(this._handlers, fn);
      return this;
   }

   /**
    * Display server error code as a console message.
    */
   // error(msg: string): (errCode: number) => void {
   //    return function(errCode: number) {
   //       //const key = statusPrefix + errCode;
   //       //const err = is.defined(text, key) ? text[key] : errCode;
   //       //log.error(msg + ":", err);
   //    };
   // }

   /**
    * Invoke every view handler.
    */
   changed() {
      this._handlers.forEach(fn => {
         fn();
      });
   }
}
