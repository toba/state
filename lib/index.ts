export { flux, StateHandler } from './hub';
export { State, StateStore, ViewHandler } from './store';
export { FluxComponent } from '../jsx/flux';
export { StateComponent } from '../jsx/stateful';
