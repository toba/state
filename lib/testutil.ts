import { State, StateStore, flux } from '../';

export interface TestState extends State {
   value: boolean;
   text?: string;
}

export class TestStore extends StateStore<TestState> {
   lastAction: number;
   lastData: any;

   constructor() {
      super({
         value: false
      });
   }

   handler(action: number, data?: any) {
      this.lastAction = action;
      this.lastData = data;
   }
}

export const makeTestStore = () => flux.subscribe(new TestStore());
