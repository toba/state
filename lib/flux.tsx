import * as React from "react";
import { State, StateStore, flux } from "./hub";

/**
 * Implement Flux pattern with a component that automatically loads its state
 * from a single "store" and dispatches events to the store. This class only
 * satisfies components that render from a single state store. More complex
 * components may need to maintain private state amalgamated from multiple
 * state stores.
 *
 * Simpler components that with unique state may use `StateComponent` instead.
 *
 * https://github.com/facebook/flux/tree/master/examples/flux-concepts
 */
export class FluxComponent<P, S extends State> extends React.Component<P, S> {
   store: StateStore<S>;

   constructor(props: P, store: StateStore<S>) {
      super(props);
      this.store = store;
      this.state = store.load();
   }

   /**
    * Subscribe component to all store changes.
    */
   componentDidMount() {
      this.store.subscribe(this.onChange.bind(this));
   }
   componentWillUnmount() {
      this.store.remove(this.onChange);
   }

   /**
    * Update component state whenever store signals a change.
    */
   onChange() {
      super.setState(this.store.state);
   }

   /**
    * Set state on store so it's the authority on state and doesn't get out of
    * sync with component state.
    */
   setState<K extends keyof S>(
      state:
         | ((prevState: Readonly<S>, props: P) => Pick<S, K> | S)
         | (Pick<S, K> | S),
      _callback?: () => any
   ): void {
      this.store.update(state);
      this.onChange();
   }

   /**
    * Emit an action to be processed by zero or more stores. If the action
    * triggers a change in the bound store then `onChange()` will be
    * called to update component state.
    *
    * If `data` will be relayed to a service call then the service must define
    * a matching `struct` to unmarshall the JSON.
    */
   emit(action: number, data?: any) {
      flux.emit(action, data);
   }
}
