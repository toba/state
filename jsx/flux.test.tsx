import * as React from "react";
import * as renderer from "react-test-renderer";
import { FluxComponent } from "../";
import { makeTestStore, TestState } from "../lib/testutil";

const testStore = makeTestStore();

class TestComponent extends FluxComponent<{}, TestState> {
   constructor(props: any) {
      super(props, testStore);
      // this.emit(Action.GetHostConfiguration);
      // this.selectTab = this.selectTab.bind(this);
   }
   render() {
      return <div />;
   }
}

test("something", () => {
   const component = renderer.create(<TestComponent />);
   const tree = component.toJSON();
   expect(tree).toMatchSnapshot();
   //expect(2).toBe(2);
});
