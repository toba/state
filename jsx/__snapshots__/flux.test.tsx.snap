// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`something 1`] = `<div />`;
